services/data/src/engine/types/Query.ts

```diff
@@ -18,9 +18,19 @@ export interface ResolvedResourceQuery extends ResourceQuery {
     params?: QueryParameters
 }
 
-export type Query = Record<string, ResourceQuery>
+export type Query<TQueryResult extends QueryResult = QueryResult> = Record<
+    keyof TQueryResult,
+    ResourceQuery
+>
 export type QueryResult = JsonMap
 
+export type QueryResultData<
+    TQuery extends Query<TQueryResultData>,
+    TQueryResultData extends QueryResult = QueryResult
```

Comment:
I think `TQueryResultData` can be removed here and use `QueryResult` directly, since this is never passed.